### PR TITLE
Support external token for bridge and add handling fee

### PR DESCRIPTION
**Pull request by rjman-ljm**
Close
  - #977

### CODE REVIEW

**Comment by rjman-ljm** [comment]
I'm working on `extends assets and membership for bridge` now, maybe review it later

File: pallets/bridge/src/lib.rs [hunk 1]
```diff
@@ -278,47 +286,81 @@ pub mod pallet {
...
+        #[transactional]
+        pub fn set_currency_fee(
+            origin: OriginFor<T>,
+            currency_id: CurrencyId,
+            new_fee: BalanceOf<T>,
+        ) -> DispatchResult {
+            Self::ensure_admin(origin)?;
+            Self::ensure_currency_registered(currency_id)?;
+
+            let asset_id = AssetIds::<T>::get(currency_id);
```
Comment (yz89) [inline pallets/bridge/src/lib.rs:345]: Why not use currency_id? It confuses that getting an asset_id by currency_id and modify the Currencies by asset_id.

File: pallets/bridge/src/lib.rs [hunk 2]
```diff
@@ -333,11 +375,28 @@ pub mod pallet {
             Self::ensure_currency_registered(currency_id)?;
 
             let asset_id = AssetIds::<T>::get(currency_id);
-            T::Assets::transfer(asset_id, &who, &Self::account_id(), amount, false)?;
+            let Currency { external, fee, .. } = Currencies::<T>::get(asset_id);
+            let total_amount = amount.checked_add(fee).ok_or(ArithmeticError::Overflow)?;
+            if external {
+                T::Assets::burn_from(asset_id, &who, total_amount)?;
+            } else {
+                T::Assets::transfer(asset_id, &who, &Self::account_id(), total_amount, false)?;
```
Comment (yz89) [inline pallets/bridge/src/lib.rs:383]: Will the relayer get the fee information from Currencies storage and reserve the fee itself?

Comment (yz89) [inline pallets/bridge/src/lib.rs:383]: Will the relayer get the fee information from Currencies storage and reserve the fee itself?

File: pallets/bridge/src/lib.rs (hunk 1 above)
Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs:345]: *Reason*
the struct of currency is complex，it's not suitable as a key of maapping

*Solutions*
1. currency_id => asset_id => currency (Just like above👆)
2. currency_id => currency (need a extra storage)

File: pallets/bridge/src/lib.rs (hunk 2 above)
Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs:383]: No, relayer only processes the transaction itself and does not make changes to the transaction information（like the fee）

Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs:383]: Fee will be reserved by `teleported`, if we want to get these reserved fees, we can sendAndSign `materialize`, just like the cross-chain extrinsics do

Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs:383]: No, relayer only processes the transaction itself and does not make changes to the transaction information. Fees will be reserved by `teleport`, if we want to get these reserved fee, we can sendAndSign the `materialize`, just like the cross-chain extrinsics do

File: pallets/bridge/src/lib.rs (hunk 1 above)
Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs:345]: This did cause a bit confusion, maybe I should add another storage for `currency_id => currency`?

File: pallets/bridge/src/lib.rs (hunk 2 above)
Comment (rjman-ljm) [reply pallets/bridge/src/lib.rs:383]: Same as above

**Review by yz89 — APPROVED** [review]

**Pull request merged**